File: sdks/python/examples/shopping_cart/checkout_log.py
"""Where the notifier records checkouts: a key value entity per cart holding when it happened."""

from __future__ import annotations

from dataclasses import dataclass

from ankka import DONE, Done, json_codec
from ankka.effects.key_value import KeyValueEffect, KeyValueReadOnlyEffect
from ankka.event_sourced_entity import command, query
from ankka.key_value_entity import KeyValueEntity


@dataclass(frozen=True)
class CheckoutRecord:
    cartId: str
    at: int = 0
    notified: bool = False


class CheckoutLog(KeyValueEntity[CheckoutRecord]):
    component_id = "checkout-log"
    state_codec = json_codec(CheckoutRecord, "checkout-record")

    def empty_state(self) -> CheckoutRecord:
        return CheckoutRecord(self.entity_id)

    @command("record")
    def record(self, at: int) -> KeyValueEffect[CheckoutRecord, Done]:
        return self.effects.update_state(CheckoutRecord(self.entity_id, at, True)).then_reply(lambda _: DONE)

    @query("get")
    def get(self) -> KeyValueReadOnlyEffect[CheckoutRecord, CheckoutRecord]:
        return self.effects.reply(self.state)

File: sdks/typescript/examples/shopping-cart/checkoutLog.ts
// Where the notifier records checkouts: a key value entity per cart holding when it happened.
import { Done, done, KeyValueEntity, command, jsonCodec, query, s, type Infer } from "ankka"

// docs:start key-value
export const CheckoutRecord = s.record("CheckoutRecord", { cartId: s.string, at: s.long, notified: s.boolean })
export type CheckoutRecord = Infer<typeof CheckoutRecord>

export class CheckoutLog extends KeyValueEntity<CheckoutRecord> {
  static readonly componentId = "checkout-log"
  static readonly state = jsonCodec(CheckoutRecord, "checkout-record")

  static readonly handlers = {
    record: command("record", s.long, Done, (log: CheckoutLog, at) => log.effects.updateState({ cartId: log.entityId, at, notified: true }).thenReply(() => done)),
    get: query("get", CheckoutRecord, (log: CheckoutLog) => log.effects.reply(log.state)),
  }

  emptyState(): CheckoutRecord {
    return { cartId: this.entityId, at: 0n, notified: false }
  }
}
// docs:end key-value
